src/librustc_mir/interpret/eval_context.rs

```diff
@@ -48,8 +48,38 @@ pub struct InterpCx<'mir, 'tcx, M: Machine<'mir, 'tcx>> {
         FxHashMap<(Ty<'tcx>, Option<ty::PolyExistentialTraitRef<'tcx>>), Pointer<M::PointerTag>>,
 }
 
+struct SpanGuard(log::span::Span);
+
+impl SpanGuard {
+    fn new() -> Self {
+        Self(log::span::Span::none())
+    }
+
+    fn enter(&mut self, span: log::span::Span) {
+        if let Some(id) = span.id() {
+            log::dispatcher::get_default(|d| {
+                d.enter(&id);
+            });
+        }
+        // This executes the destructor on the previous instance of `SpanGuard`, ensuring that
+        // we never enter or exit more spans than vice versa. Unless you `mem::leak`, then we
+        // can't protect the tracing stack, but that'll just lead to weird logging, no actual
+        // problems.
+        *self = Self(span);
+    }
+}
+
+impl Drop for SpanGuard {
+    fn drop(&mut self) {
+        if let Some(id) = self.0.id() {
+            log::dispatcher::get_default(|d| {
+                d.exit(&id);
+            });
+        }
+    }
+}
+
 /// A stack frame.
-#[derive(Clone)]
```

Comment:
`tracing::Span` is `Clone`, so I think this type _could_ continue to derive clone...

Reply:
I agree with this diff though, there's no reason we should clone a stack frame.

Reply:
yea this is more of a drive-by fix